### PR TITLE
iio: Add new nodes for light intensity sensors

### PR DESCRIPTION
**Pull request by yongli3**
To support intensity_both/intensity_ir data reading in some light sensors

Signed-off-by: Yong Li <yong.b.li@intel.com>

### CODE REVIEW

**Comment by yongli3** [comment]
@edersondisouza @fulong82 @laykuanloon please help to review, this PR is used for issue #2244

File: src/modules/flow/iio/nodes.c [hunk 1]
```diff
@@ -943,6 +943,154 @@ light_open(struct sol_flow_node *node, void *data, const struct sol_flow_node_op
 }
 
 static bool
+intensity_both_create_channels(struct iio_double_data *mdata, int device_id)
+{
+    mdata->iio_base.device = sol_iio_open(device_id, &mdata->iio_base.config);
+    SOL_NULL_CHECK(mdata->iio_base.device, false);
+
+    mdata->channel_val = iio_add_channel(mdata->scale, mdata->offset, "in_intensity_both", &mdata->iio_base);
```
Comment (yongli3) [inline src/modules/flow/iio/nodes.c]: @fulong82 Thanks for your test. I am using the TSL2563 as below. Please let me know which the sensor name you are testing, regarding other sensors RPR0521/LTR501/APDS9300/HID ALS, do you know if they are using the in_intensity_offset or in_intensity_both_offset ?
 cat name
tsl2563
root@intel-quark:/sys/bus/iio/devices/iio:device1# ls -l
-r--r--r--    1 root     root          4096 Sep 27 15:27 dev
drwxr-xr-x    2 root     root             0 Sep 27 15:27 events
-rw-r--r--    1 root     root          4096 Sep 27 15:27 in_illuminance0_input
-rw-r--r--    1 root     root          4096 Sep 27 15:27 in_intensity_both_calibscale
-rw-r--r--    1 root     root          4096 Sep 27 15:27 in_intensity_both_raw
-rw-r--r--    1 root     root          4096 Sep 27 15:27 in_intensity_ir_calibscale
-rw-r--r--    1 root     root          4096 Sep 27 15:27 in_intensity_ir_raw
-r--r--r--    1 root     root          4096 Sep 27 15:27 name
drwxr-xr-x    2 root     root             0 Sep 27 15:27 power
lrwxrwxrwx    1 root     root             0 Sep 27 15:27 subsystem -> ../../../../../../../bus/iio
-rw-r--r--    1 root     root          4096 Sep 27 15:27 uevent

Comment (yongli3) [inline src/modules/flow/iio/nodes.c]: @fulong82 I updated the PR, please help to test if the scale/offset is correct

Comment (fulong82) [inline src/modules/flow/iio/nodes.c]: @yongli3
Tested. It is working now. Thx.

File: src/lib/io/sol-iio.c [hunk 2]
```diff
@@ -544,12 +544,27 @@ channel_get_pure_name(struct sol_iio_channel *channel)
...
     size_t channel_name_len;
     char *channel_pure_name;
+    char *channel_name_suffix;
     bool modified = false;
 
     channel_name_len = strlen(channel->name);
-    if (channel_name_len > 2) {
-        char *channel_name_suffix;
+    if (channel_name_len > 5) {
+        channel_name_suffix = channel->name + channel_name_len - 5;
```
Comment (edersondisouza) [inline src/lib/io/sol-iio.c]: According to kernel doc, `Documentation/ABI/testing/sysfs-bus-iio`, it's possible to have a channel like `in_intensityY_both_raw`, where Y is an index number. So I'd say that you have to take this into consideration. An idea would be doing what you do now, stripping suffix 'ir' and 'both' (and maybe 'uv') then letting code below process the Y part.
Note that Soletta also has a `strendswith` macro, that may help you to check a string suffix.

File: src/lib/io/sol-iio.c [hunk 3]
```diff
@@ -544,12 +544,27 @@ channel_get_pure_name(struct sol_iio_channel *channel)
 {
```
Comment (edersondisouza) [inline src/lib/io/sol-iio.c]: Please, also update this function commentary, to include this 'both' and 'ir' thing.
That said, shouldn't you also include 'uv' suffix? It appears on kernel doc...

File: src/modules/flow/iio/iio.json [hunk 4]
```diff
@@ -945,6 +945,210 @@
...
+             "description": "Range of output packet. Usually, is the output range of used sensor. It'll be the 'min' and 'max' fields of drange packets sent on OUT port.",
+             "name": "out_range"
+         }
+       ],
+       "version": 1
+      },
+      "out_ports": [
+        {
+         "data_type": "float",
+         "description": "Light data read in lux.",
```
Comment (edersondisouza) [inline src/modules/flow/iio/iio.json]: I had understood that this data is unit-less, isn't it?

File: src/modules/flow/iio/iio.json [hunk 5]
```diff
@@ -945,6 +945,210 @@
...
```
Comment (edersondisouza) [inline src/modules/flow/iio/iio.json]: Ditto.

File: src/modules/flow/iio/iio.json (hunk 4 above)
Comment (yongli3) [inline src/modules/flow/iio/iio.json]: @edersondisouza some light sensors provide the data in lux, some are raw data. For raw data we will use the scale factor to process them. For "unit-less", do you mean we need to change it to  
"Light data read in lux(unit-lesss)" ?

File: src/lib/io/sol-iio.c (hunk 2 above)
Comment (yongli3) [inline src/lib/io/sol-iio.c]: Please search the "in_intensity_both" in nodes.c, in order to support the Y(a number) part, we need to add the channel_id(multi channel) logic ?

File: src/modules/flow/iio/iio.json (hunk 4 above)
Comment (edersondisouza) [inline src/modules/flow/iio/iio.json]: That's good.

File: src/lib/io/sol-iio.c (hunk 2 above)
Comment (edersondisouza) [inline src/lib/io/sol-iio.c]: No, it's not multi-channel logic. Look in this function, but some lines below. The job of stripping the Y component is already done. Just change the flow so it goes there after you stripped 'both' or 'ir' part.

File: src/modules/flow/iio/iio.json (hunk 5 above)
Comment (yongli3) [inline src/modules/flow/iio/iio.json]: Done

**Comment by yongli3** [comment]
@edersondisouza @fulong82 I updated the PR, please help to review/test it

File: src/modules/flow/iio/nodes.c [hunk 6]
```diff
@@ -959,6 +959,160 @@ light_open(struct sol_flow_node *node, void *data, const struct sol_flow_node_op
 }
 
 static bool
+intensity_both_create_channels(struct iio_double_data *mdata, int device_id)
+{
+    mdata->iio_base.device = sol_iio_open(device_id, &mdata->iio_base.config);
+    SOL_NULL_CHECK(mdata->iio_base.device, false);
+
+    mdata->channel_val = iio_add_channel(mdata->scale, mdata->offset, "in_intensity0_both", &mdata->iio_base);
```
Comment (edersondisouza) [inline src/modules/flow/iio/nodes.c]: I think that as multiple channels support is already in, you can rebase this patch on top of master and make use of that, instead of this `in_intensity0_both` and `in_intensity_both` trick.

File: src/modules/flow/iio/nodes.c [hunk 7]
```diff
@@ -959,6 +959,160 @@ light_open(struct sol_flow_node *node, void *data, const struct sol_flow_node_op
...
+
+}
+
+static bool
+intensity_ir_create_channels(struct iio_double_data *mdata, int device_id)
+{
+    mdata->iio_base.device = sol_iio_open(device_id, &mdata->iio_base.config);
+    SOL_NULL_CHECK(mdata->iio_base.device, false);
+
+    mdata->channel_val = iio_add_channel(mdata->scale, mdata->offset, "in_intensity0_ir", &mdata->iio_base);
```
Comment (edersondisouza) [inline src/modules/flow/iio/nodes.c]: Ditto.

File: src/lib/io/sol-iio.c [hunk 8]
```diff
@@ -673,15 +696,15 @@ iio_set_channel_scale(struct sol_iio_channel *channel, double scale)
     /* If failed, try channel pure name */
     pure_name = channel_get_pure_name(channel);
     if (pure_name && craft_filename_path(path, sizeof(path), CHANNEL_SCALE_PATH,
-        channel->device->device_id, channel->name)) {
+        channel->device->device_id, pure_name)) {
 
-        result = (sol_util_write_file(path, "%lf", scale) > 0);
+        result = (sol_util_write_file(path, "%.9lf", scale) > 0);
```
Comment (edersondisouza) [inline src/lib/io/sol-iio.c]: Why this change?

File: src/lib/io/sol-iio.c [hunk 9]
```diff
@@ -595,22 +615,25 @@ channel_get_pure_name(struct sol_iio_channel *channel)
...
+            }
 
             free(original_channel_pure_name);
-            return NULL;
+            return internsity_channel_name;
         }
     }
 
-    return NULL;
+    return internsity_channel_name;
```
Comment (edersondisouza) [inline src/lib/io/sol-iio.c]: I guess that this is over complicated...
A simple approach to use would be (highlevel pseudocode):
  - copy `channel->name` to a `channel_name[NAME_MAX]` buffer
  - `strendswith(channel_name, "_both") || strendswith(channel_name, "_ir")` ->strip from channel_name, doing `channel_name[strlen(channel_name) - 5 (or 3)] = '\0'`
  - let current code act, but on `channel_name` instead of `channel->name`
On last step, you could actually do the same for `_x`, `_y` and `_z` stripping, so code would be uniform, but it isn't a must for now.
Any thoughts?

File: src/lib/io/sol-iio.c [hunk 10]
```diff
@@ -709,7 +732,7 @@ iio_set_channel_offset(struct sol_iio_channel *channel, int offset)
     /* If failed, try channel pure name */
     pure_name = channel_get_pure_name(channel);
     if (pure_name && craft_filename_path(path, sizeof(path), CHANNEL_OFFSET_PATH,
-        channel->device->device_id, channel->name)) {
+        channel->device->device_id, pure_name)) {
```
Comment (edersondisouza) [inline src/lib/io/sol-iio.c]: Nice catch... guess this was wrong for a long time :cold_sweat:

File: src/lib/io/sol-iio.c (hunk 9 above)
Comment (yongli3) [inline src/lib/io/sol-iio.c]: I prefer to keep the current complicated implement, in order to resolve the issue #2244. I will refactor this function in the next PR

Comment (edersondisouza) [inline src/lib/io/sol-iio.c]: But alternative is easier, actually.

File: src/lib/io/sol-iio.c (hunk 8 above)
Comment (edersondisouza) [inline src/lib/io/sol-iio.c]: How about `%g`?

Comment (edersondisouza) [inline src/lib/io/sol-iio.c]: Ah, true. I was thinking about log - it would be easier to see.

File: src/modules/flow/iio/nodes.c (hunk 6 above)
Comment (yongli3) [inline src/modules/flow/iio/nodes.c]: @edersondisouza The current GEN_CHANNEL_NAME macro is used to append a "%d" at the end of the string. But the two in_intensity0_both and in_intensity0_ir have a different format(%d is in hte middle of the string), any thoughts/suggestions about the change?

Comment (edersondisouza) [inline src/modules/flow/iio/nodes.c]: Don't use the macro, treat them differently - they are the exception, macro is for generic case.

Comment (yongli3) [inline src/modules/flow/iio/nodes.c]: Done

File: src/lib/io/sol-iio.c (hunk 9 above)
Comment (yongli3) [inline src/lib/io/sol-iio.c]: @edersondisouza I changed this function, please help to review

File: src/lib/io/sol-iio.c (hunk 2 above)
Comment (yongli3) [inline src/lib/io/sol-iio.c]: Done

File: src/modules/flow/iio/nodes.c [hunk 11]
```diff
@@ -959,6 +959,178 @@ light_open(struct sol_flow_node *node, void *data, const struct sol_flow_node_op
...
+
+    mdata->iio_base.device = sol_iio_open(device_id, &mdata->iio_base.config);
+    SOL_NULL_CHECK(mdata->iio_base.device, false);
+
+    if (channel_id >= 0) {
+        ret = snprintf(channel_name, sizeof(channel_name), "in_intensity%d_both", channel_id);
+        SOL_INT_CHECK_GOTO(ret, >= (int)sizeof(channel_name), error);
+        SOL_INT_CHECK_GOTO(ret, < 0, error);
+    } else {
+        strncpy(channel_name, "in_intensity_both", sizeof(channel_name) - 1);
```
Comment (edersondisouza) [inline src/modules/flow/iio/nodes.c]: Hint: as you are copying a constant into buffer, no strong need for `strncpy` and '\0' worries. But it doesn't hurt, either.

File: src/lib/io/sol-iio.c [hunk 12]
```diff
@@ -610,6 +628,11 @@ channel_get_pure_name(struct sol_iio_channel *channel)
         }
     }
 
+    if (modified) {
+        channel_pure_name = strndup(channel_name, channel_name_len);
```
Comment (edersondisouza) [inline src/lib/io/sol-iio.c]: `strndup` may fail. I know that there's a non-verified call above. It should be fixed too. It maybe on another patch, though.

**Comment by edersondisouza** [comment]
Integrated, tks!

**Comment by yongli3** [comment]
Thanks @edersondisouza I have tested it on my boards using TSL2563, HMC5843 and L3G4200D. I will submit another PR for the strndup checking. @fulong82 please help to re-test it using your HID sensors

**Comment by fulong82** [comment]
@yongli3,

Tested working. Thanks. 👍

**Comment by fulong82** [comment]
@laykuanloon & @yongli3 

Btw, i saw this warning message when using the latest code base. Do we need to suppress it as not all iio device will have mount_matrix especially  ambient light sensor.

```
# ../runfbsh als.fbp sol-flow-hid.json
WRN: ./src/lib/io/sol-iio.c:770 get_mount_matrix() Could not find mount_matrix for device2
illuminance = 103.583000 (float range) Lux
illuminance = 104.473000 (float range) Lux
illuminance = 103.880000 (float range) Lux
```

**Comment by yongli3** [comment]
@fulong82 Thanks for your quick response! Regarding this WRN message, I think we can change it to DBG, any comments/suggestions?

**Comment by fulong82** [comment]
@yongli3 I am fine for SOL_DBG. :-)

@laykuanloon What do you think? Since you are the requester for it.

**Comment by laykuanloon** [comment]
@fulong82 @yongli3 
I am ok.
